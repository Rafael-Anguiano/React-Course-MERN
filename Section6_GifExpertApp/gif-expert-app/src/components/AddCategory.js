import React, { useState } from 'react'
import PropTypes from 'prop-types'

const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('')

    const handleInput = (e) => {
        setInputValue(e.target.value)
        // console.log('HandleInputChange called');
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        // console.log('HandleSubmit', inputValue);
        if(inputValue.trim().length > 2 ){
            setCategories( cats => [inputValue, ...cats])
            setInputValue("");
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <input 
                type="text"
                value={inputValue}
                onChange={handleInput}
            />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

export default AddCategory;