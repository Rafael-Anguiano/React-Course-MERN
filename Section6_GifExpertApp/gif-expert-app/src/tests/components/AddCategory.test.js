import React from 'react'
import { shallow } from "enzyme"
import AddCategory from "../../components/AddCategory"
import '@testing-library/jest-dom'


describe('Tests in file AddCategory', () => {

    const setCategories = jest.fn();
    let wrapper;
    
    beforeEach(()=>{
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={ setCategories } />)
    })

    test('should match with snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    test('should change textbox', () => {
        const input = wrapper.find('input');
        const value = 'Hello World'

        input.simulate('change', { target: {value} } );

    })

    test('should not post information on submit', () => {
        wrapper.find('form').simulate('submit', {preventDefault(){}} )
        expect(setCategories).not.toHaveBeenCalled()
    })
    
    test('should call setCategories and clear textInput', () => {
        // 1. Simulate Input Change
        const input = wrapper.find('input');
        const value = 'Hello World'
        input.simulate('change', { target: {value} } );
        
        // 2. Simulate Form submit
        wrapper.find('form').simulate('submit', {preventDefault(){}} )
        expect(setCategories).toHaveBeenCalled()
        expect(setCategories).toHaveBeenCalledTimes(1)
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function))
        
        // 3. Validate reset of the Input
        expect( input.prop('value') ).toBe('')

    })
    
})
