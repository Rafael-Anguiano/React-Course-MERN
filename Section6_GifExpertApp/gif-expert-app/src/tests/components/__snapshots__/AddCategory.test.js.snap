// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in file AddCategory should match with snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
